lib/fpm/package.rb

```diff
@@ -249,8 +257,20 @@ def output(path)
                                   "creating #{self.type} packages")
   end # def output
 
+  def temporary_directory(path)
+    if attributes[:source_date_epoch].nil?
+      return Stud::Temporary.directory("package-#{type}-staging")
+    else
+      # Choose a predictable build path; see https://reproducible-builds.org/docs/build-path/
+      root = ENV["TMP"] || ENV["TMPDIR"] || ENV["TEMP"] || "/tmp"
```

Comment:
If you run the command "fpm -s gem -t deb json" twice in a row, it generates two different debs.
In the reproducible builds use case, we want it to generate two identical debs, regardless of who
runs the build when or where, as long as the OS is the same release / version / architecture. 
(Otherwise e.g. clicking 'Force Rebuild' in buildbot generates failed uploads because reprepro rejects
the new deb with the same version and different checksum.)

One particular problem is that the C toolchain often embeds the absolute path to the build
directory in the binary, say, in the DW_AT_comp_dir field.
Until we teach the compiler to obey BUILD_PATH_PREFIX_MAP,
we can work around this by using a deterministic build path
( see https://reproducible-builds.org/docs/build-path/ ).

So, this has nothing to do with leaving a clean build machine after running fpm, but everything to do with bit-for-bit identical output packages given bit-for-bit identical input packages.